src/main/java/io/cdap/plugin/gcp/bigquery/connector/BigQueryConnector.java

```diff
@@ -0,0 +1,370 @@
+      .build();
+  }
+
+  private String[] parsePath(String originalPath) {
+    String path = originalPath;
+    if (originalPath.startsWith("/")) {
+      path = originalPath.substring(1);
+    }
+    String[] parts = path.split("/");
+    // path should contains at most two part : dataset and table
+    if (parts.length > 2) {
+      throw new IllegalArgumentException(
+        String.format("Path %s is invalid. Path should at most contain two parts.", originalPath));
+    }
+    return parts;
+  }
+
+  private BrowseDetail listTables(BigQuery bigQuery, String dataset, Integer limit) {
+    int countLimit = limit == null ? Integer.MAX_VALUE : limit;
+    int count = 0;
+    BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+    DatasetId datasetId = DatasetId.of(dataset);
+    Page<Table> tablePage = bigQuery.listTables(datasetId);
+    String parentPath = "/" + dataset + "/";
+    for (Table table : tablePage.iterateAll()) {
+      String name = table.getTableId().getTable();
+      browseDetailBuilder
+        .addEntity(BrowseEntity.builder(name, parentPath + name, ENTITY_TYPE_TABLE).canSample(true).build());
+      count++;
+      if (count == countLimit) {
+        break;
+      }
+    }
+    return browseDetailBuilder.setTotalCount(count).build();
+  }
+
+  private BrowseDetail listDatasets(BigQuery bigQuery, Integer limit) {
+    Page<Dataset> datasetPage = bigQuery.listDatasets(BigQuery.DatasetListOption.all());
+    int countLimit = limit == null ? Integer.MAX_VALUE : limit;
+    int count = 0;
+    BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+    for (Dataset dataset : datasetPage.iterateAll()) {
+      String name = dataset.getDatasetId().getDataset();
+      browseDetailBuilder
+        .addEntity(BrowseEntity.builder(name, "/" + name, ENTITY_TYPE_DATASET).canBrowse(true).build());
+      count++;
+      if (count == countLimit) {
+        break;
+      }
+    }
+    return browseDetailBuilder.setTotalCount(count).build();
+  }
+
+  private BigQuery getBigQuery() throws IOException {
+    GoogleCredentials credentials = null;
+    //validate service account
+    if (config.isServiceAccountJson() || config.getServiceAccountFilePath() != null) {
+      credentials =
+        GCPUtils.loadServiceAccountCredentials(config.getServiceAccount(), config.isServiceAccountFilePath());
+    }
+    ;
+    return GCPUtils.getBigQuery(config.getDatasetProject(), credentials);
+  }
+
+  private List<StructuredRecord> getData(BigQuery bigQuery, String dataset, String table, int limit)
+    throws IOException {
+    String query = String.format("SELECT * FROM `%s.%s.%s` LIMIT %d", config.getDatasetProject(), dataset, table,
+      limit);
+    QueryJobConfiguration queryConfig = QueryJobConfiguration.newBuilder(query).build();
+    JobId jobId = JobId.of(UUID.randomUUID().toString());
+    Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
```

Comment:
I believe this means the job is always in the default location (US). Have you tested on tables in different locations?

Reply:
good point. It's the original logic from Wrangler.
And I just tested , it can query tables in europe.